File: scripts/validate.mjs
#!/usr/bin/env node
/**
 * Prüft alle `src/data/*.json` gegen `schema/medium.schema.json`.
 * Exit-Code 1 bei Fehlern.
 *
 * Geprüft wird dreierlei — genau wie zuvor in `scripts/validate.py`:
 *   1. Jeder Eintrag erfüllt das JSON Schema.
 *   2. Die Zahl `anzahl` stimmt mit der Zahl der Einträge überein.
 *   3. Keine `id` kommt zweimal vor.
 *
 * Warum in Node und nicht in Python: Diese Prüfung hängt als `prebuild` vor jedem
 * Build und lief damit auch auf dem Deploy-Server. Dort einen Python-Interpreter samt
 * `jsonschema` vorauszusetzen, hieß, dass ein fehlendes `pip3` das Veröffentlichen
 * blockiert — für eine Prüfung, die keine Python-Eigenschaft braucht. Node ist
 * ohnehin da, denn ohne Node gibt es keinen Build.
 *
 * `scripts/validate.py` bleibt für die lokale Arbeit erhalten und prüft dasselbe.
 * Die GitHub Action lässt beide laufen — wenn sie je auseinanderlaufen, fällt es
 * dort auf und nicht erst im Bestand.
 *
 * Punkt 1 steckt nicht mehr in diesem Skript, sondern in `src/lib/pruefung.ts`:
 * Dieselbe Funktion prüft später einen einzelnen Eintrag, bevor die Verwaltung ihn
 * ins Repository schreibt. Eine Prüfung, zwei Aufrufer — damit kann gar nicht erst
 * entstehen, was der Absatz darüber für validate.py befürchtet. Punkt 2 und 3 bleiben
 * hier: Sie gelten für eine ganze Datei, nicht für einen einzelnen Eintrag.
 *
 * Das Skript lädt mit `pruefung.ts` erstmals ein TypeScript-Modul und wird deshalb in
 * `package.json` mit `--experimental-strip-types` gestartet. Node kann das ab 22.18
 * von allein; die Flagge deckt die älteren 22er mit ab, die auf einem Deploy-Server
 * stehen können, und `--disable-warning=ExperimentalWarning` sorgt dafür, dass die
 * Ausgabe dabei auf jeder Node-Fassung dieselbe bleibt.
 */
import { readFileSync, readdirSync } from 'node:fs';
import { dirname, join, resolve } from 'node:path';
import { fileURLToPath } from 'node:url';
import { pruefeMedium } from '../src/lib/pruefung.ts';

/** Ab so vielen Fehlern wird die Ausgabe abgeschnitten — sonst scrollt sie weg. */
const MAX_MELDUNGEN = 25;

// Dieses Skript liegt in scripts/, die Daten liegen darüber. Das Schema muss hier
// nicht mehr gesucht werden; pruefung.ts bringt es mit.
const wurzel = resolve(dirname(fileURLToPath(import.meta.url)), '..');
const datenOrdner = join(wurzel, 'src', 'data');

const dateien = readdirSync(datenOrdner)
  .filter((name) => name.endsWith('.json'))
  // Dateien mit führendem Unterstrich sind Arbeitsmaterial, kein Bestand.
  .filter((name) => !name.startsWith('_'))
  .sort();

let fehlerzahl = 0;
let gesamt = 0;
/** id → wie oft sie vorkommt. */
const ids = new Map();

function melde(text) {
  fehlerzahl += 1;
  if (fehlerzahl <= MAX_MELDUNGEN) console.log(text);
  else if (fehlerzahl === MAX_MELDUNGEN + 1) console.log('... (weitere Fehler unterdrückt)');
}

for (const name of dateien) {
  const inhalt = JSON.parse(readFileSync(join(datenOrdner, name), 'utf8'));
  const items = inhalt.items ?? [];
  gesamt += items.length;

  if (inhalt.anzahl !== items.length) {
    melde(`${name}: anzahl=${inhalt.anzahl} stimmt nicht mit ${items.length} Einträgen überein`);
  }

  items.forEach((eintrag, index) => {
    ids.set(eintrag.id, (ids.get(eintrag.id) ?? 0) + 1);

    const ergebnis = pruefeMedium(eintrag);
    if (!ergebnis.gueltig) {
      for (const zeile of ergebnis.fehler) {
        melde(`${name}[${index}] ${eintrag.id}: ${zeile}`);
      }
    }
  });

  console.log(`${name.padEnd(28)} ${String(items.length).padStart(5)} Einträge`);
}

for (const [id, anzahl] of ids) {
  if (anzahl > 1) melde(`DOPPELTE ID: ${id} (${anzahl}x)`);
}

console.log(`\ngesamt ${gesamt} Einträge, ${fehlerzahl} Fehler`);

if (fehlerzahl > 0) {
  console.error(
    '\nDie Datenprüfung ist fehlgeschlagen. Der Build wurde abgebrochen,\n' +
      'damit fehlerhafte Daten nicht veröffentlicht werden.\n',
  );
}

process.exit(fehlerzahl > 0 ? 1 : 0);

File: scripts/validate.py
#!/usr/bin/env python3
"""Prüft alle src/data/*.json gegen schema/medium.schema.json. Exit-Code 1 bei Fehlern.

NICHT Teil des Builds. Die Prüfung, die vor jedem Build läuft, steckt in
scripts/validate.mjs und braucht nur Node — damit auf dem Deploy-Server kein Python
vorhanden sein muss. Dieses Skript bleibt für die lokale Arbeit erhalten und prüft
dasselbe: Schemakonformität, doppelte ids, und dass `anzahl` zur Zahl der Einträge
passt.

Aufruf: `npm run validate:py` (oder `python scripts/validate.py`).
Braucht `pip install -r requirements.txt`.
"""
import json, sys, glob, os
from collections import Counter
from jsonschema import Draft202012Validator

# Umlaute auch auf Windows-Konsolen ausgeben können, ohne UnicodeEncodeError.
if hasattr(sys.stdout, 'reconfigure'):
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')

# Dieses Skript liegt in scripts/, die Daten und das Schema liegen im Projektwurzelverzeichnis.
base = os.path.dirname(os.path.dirname(os.path.abspath(__file__)))
schema = json.load(open(os.path.join(base,'schema/medium.schema.json'), encoding='utf-8'))
v = Draft202012Validator(schema)
fehler = 0; ids = Counter(); gesamt = 0

for pfad in sorted(glob.glob(os.path.join(base,'src/data/*.json'))):
    name = os.path.basename(pfad)
    if name.startswith('_'): continue
    d = json.load(open(pfad, encoding='utf-8'))
    items = d.get('items', [])
    gesamt += len(items)
    if d.get('anzahl') != len(items):
        print(f'{name}: anzahl={d.get("anzahl")} stimmt nicht mit {len(items)} Einträgen überein'); fehler += 1
    for i, it in enumerate(items):
        ids[it.get('id')] += 1
        for e in v.iter_errors(it):
            fehler += 1
            if fehler <= 25:
                print(f'{name}[{i}] {it.get("id")}: {".".join(map(str,e.path))} {e.message}')
            elif fehler == 26:
                print('... (weitere Fehler unterdrückt)')
    print(f'{name:28} {len(items):5} Einträge')

for k, n in ids.items():
    if n > 1: print(f'DOPPELTE ID: {k} ({n}x)'); fehler += 1

print(f'\ngesamt {gesamt} Einträge, {fehler} Fehler')
sys.exit(1 if fehler else 0)
